### ISSUE TITLE
set and get object properties

### ISSUE THREAD

**Issue by fabianmoronzirfas** [basiljs/basil.js#43]
There should be an easy way to set and get properties of an object. it makes no sense to implement a own getter and setter for all the things you can do in InDesign but there needs to be a way to set things like the pageWidth with a simple call. e.g.  

```
b.property(b.doc(), "pageWidth", value);
```

and

```
b.property(b.doc(), "pageWidth");
```

The biggest challenge is the nesting of these properties. The InDesign API call is

```
app.activeDocument.documentPreferences.pageWidth
```

**Comment by b-g** [comment]
Good call! I would prefer simply `b.property()` :)

**Comment by ffd8** [comment]
Bringing this back up, since I agree there's some great potential here compared to many individual functions as proposed in #121  – particularly if it's possible to combine some nice transformations within the same function. I'm thinking, similar to `b.typo(item, "property", "value");` but for dealing with objects/items/..? ie. something that's been missing is an easy way to rotate something after drawing it. I guess it depends on how flexible the item thrown in can be and like @fabianmoronzirfas mentioned, nesting the various inputs for their exact InDesign coded needs... but it's something big missing.. the same b.typo() magic function for objects. 

Semantically, possible options would be:
- `b.property();` [what it does, but not specific to objects]
- `b.item();` [sounds closest to b.typo.. but is type also item?]
- `b.object();` [2nd closest to b.typo]
- `b.transform();` [taken.. but fitting]
- ...?

Amazing if it could act as a getter + setter, depending on if that 3rd parameter was included, but setting would be a priority.

**Comment by ffd8** [comment]
Thinking more about this.. and trying to figure out wording that would treat items/images like text/typo is being dealt with:

below using such a format: 
- b.generatorFunction() / b.transformFunction()

### Typography
- b.text() / b.typo() » [clear and exists]

### Items
- b.rect() / b.item() » [item is semi generic]
- b.rect() / b.shape() » [limited to forms]
- b.rect() / b.form() » [inspired by previous commentary]

### Images
- b.image() / b.image() » [damn!]
- b.image() / b.modImage() » [modX could be short format for all types]
- b.image() / b.imagery() » [slightly confusing, but text -> typo.graphy]

...??

**Comment by trych** [comment]
My 2 cents:

While I agree that `b.typo()` is a nice function, I am not sure how much sense it makes to introduce some equivalent to transform objects. The thing with `b.typo()` is that it simply gives you access for getting and setting text object properties as they are described in InDesign's object model.

However, you can't really do the same with (page)items, as InDesign's object model offers no properties like `width`, `height`, `position`, `size`, `x` and `y`. That's why it does make sense to have convenience functions like

```js
b.itemHeight()
b.itemWidth()
b.itemPosition()
b.itemSize()
b.itemX()
b.itemY()
```

I think to have a `b.property()` method totally makes sense.

But introducing methods like `b.imagery()` that either would do just the same as `b.property()` or use some properties that do not even exist in InDesign's object model (`width`, `position`), would be really confusing. Especially if there are several of those methods (`b.imagery()`, `b.shape()`, `b.item()`). IMHO this would be of little benefit compared to the `b.itemXXX()` functions we do have now.

**Comment by trych** [comment]
Giving this some more thought and doing some sort of 180° turn here (maybe I should properly wake up before I post stuff ... ;) ), how about a method `b.transforms()` (as in "get transforms"/"set transforms" and also as in not to clash with `b.transform()`)?

This method could be used transform all sort of pageItems (no matter if shape, image, graphic or textFrame) using a predefined list of non-native properties like:

```js
height
width
size // -> [width, height]
x
y
position // -> [x, y]
rotation
scale // -> when setting use current scale as reference, when getting always return 1
shear
```

It could be used very much like `b.typo()` then, except it would not have native InDesign properties, but the predefined properties that we made up:

`b.transforms(pageItem, "width", 24);`

And maybe we could allow an optional anchorPoint argument:

`b.transforms(pageItem, "width", 24, b.BOTTOM_RIGHT);`

I think as long as we document this well and make a clear, pre-defined list that is easy to find, then it will be no less confusing than `b.itemXXX()`.

**Comment by ffd8** [comment]
Exactly. This is what was intended with the initial discussion behind a b.property(item, "property", "value") function– something that can fit just about any/all pageItems. Above with the various categories, was just a search for the right wording if, say, image had to be separated from pageItem. Nevertheless, only one was being suggested.. the question is how it should be called. 

Sooo yeah, back to that, 'transforms' is awfully close to 'transform' = could be confusing. Maybe `b.property()` is best or `b.modify()`. Small details, for how the function would actually work, but important for making it obvious and easy to find.

**Comment by trych** [comment]
`b.property()` is not a good choice IMHO as it implies that you can set the actual properties of an object, which is not the case here as we introduce a list of "fake" properties. I actually thought `b.property()` would be a function to allow to set all sorts of properties to all sorts of objects (not only `pageItems`, but also `paragraphStyles`, `colors` etc.)

A name like `b.modify()` does not really work either as that implies it is a setter only ("modify the parameters"). I actually think `b.transforms()` describes pretty well, what you can get and set with the function (get transforms/set transforms).

> Sooo yeah, back to that, 'transforms' is awfully close to 'transform' = could be confusing.

Yes, the same problem that we have with `b.line()` and `b.lines()`. 😜 But one side question: Do we actually need the current `b.transform()`? Just looked it up, neither Processing nor P5.js have a `transform()` method.

**Comment by ffd8** [comment]
> But one side question: Do we actually need the current b.transform()?

True true. To my understanding (never had to use it myself) it allows a matrix to be applied to an existing object, rather than pushing/popping or setting a new one for any following objects. While it kills someones rare project that used it, it would be a great function name to take over for this and could perhaps just be used within: `b.transform(item, "matrix", matrix);`

**Comment by trych** [comment]
> True true. To my understanding (never had to use it myself) it allows a matrix to be applied to an existing object, rather than pushing/popping or setting a new one for any following objects.

Never used it either, but is this actually a scenario that realistically ever happens? To first create an object and then at a later point apply a matrix? I think either you apply a matrix right away (well, as it happens automatically) or if you later want to transform an object, then you would want to do that with absolute/relative values, not with a matrix, I think. So, I would vote to get rid of it.

> `b.transform(item, "matrix", matrix);` <- or whatever the exact object model wording is..

I still think `b.transforms()` is a better name. `transform` is the verb and so it sounds like you can only transform an object and not get transform properties back. `b.transforms()` would suggest to get and set transforms and to handle all the transforms of the pageItem.

Compare this naming scheme:

`b.translate()` --> verb, only allows to translate something, not to receive a translation
`b.rotate()` --> verb, only allows to rotate something, not to receive a rotation

Hence

`b.transform()` --> verb, would logically suggest to only allow to transform something and not receive transform properties. The naming scheme also suggests it belongs to `b.translate()`, `b.rotate()` and so on, which it does not, according to the ideas above, so I think a name like

`b.transforms()`

sets it apart, makes it more clear that you can get/set transforms and handle the transforms of an object and in case we discontinue `b.transform()`, then there should be also no confusion.

**Comment by ffd8** [comment]
Just about anything is fine and curious what other opinions are. 
Adding an additional short name possibility to the mix, how about:

`b.typo()` = 4 chars for type
`b.prop()` = 4 chars for item?

**Comment by trych** [comment]
Ok, I have a feeling, we are talking about two different things here now and should take a step back. What is actually suggested here now? I thought we are talking about two new functions now:

1. a new function `b.property()` (name yet to be decided) to universally get and set props of all sort of objects, also non-pageItem-objects like paragraphStyles, settings etc.

2. a new function `b.transforms()` (name yet to be decided) to only work on "made up" transform properties (`width`, `position` etc.) of *pageItems only* and to (probably) ultimately replace functions like `b.itemWidth()` etc.

Is that correct? Or did I misunderstand something?

**Comment by trych** [comment]
I wonder if it would be of help to have another Skype call? As I remember you guys had one in spring, to discuss the general direction of basil. The project has moved forward since then and at the moment there are a lot of ideas and proposals floating around, so I would find it very helpful to discuss things "in person" at some point.

@b-g, @ffd8, @fabianmoronzirfas, what do you think?

**Comment by b-g** [comment]
Yes let's have a call ... see my email.

**Comment by ffd8** [comment]
Quick note here from our Skype, we decided on the basic wrapper of: 
`b.property(obj, "property", value);` 
– acting as a basic wrapper for any of the settings found in the Jongware docs. This is something students have recently struggled with (and is unique to InDesign vs Processing.. styling things after they're made). 

At the moment it's like so:

PRE-STYLING
`b.fill(255, 0, 0);`
`b.rect(50, 50, 50, 50);`

POST-STYLING
`var myRect = b.rect(50, 50, 50, 50);`
`myRect.fillColor = b.color(255, 0, 0);`

Changes from Processing/basil.js coding to the normal JS style... So we should start off with that most basic wrapper that just keeps the notation the same, so above's POST-STYLING becomes:

`b.property(myRect, "fillColor", b.color(255, 0, 0));`

Which properties are allowed... all dependent on the object being fed in... researching the Jongware pages... and cheat-sheets (like we have for b.typo)... (missing) tutorial/video for walking a beginner through how to research those docs and find ones way to that property setting.

**Comment by trych** [comment]
Was implemented in #278, closing.
